### ISSUE TITLE
Inherits vs Implements

### ISSUE THREAD

**Issue by ilyapuchka** [krzysztofzablocki/Insanity#8]
Currently `implements` is just an alias for `inherits`. But wouldn't it align more with language concepts of implementing some protocol and inheriting from some type if `inherits` will only contain inherited type (which can be only one and only for classes) and `implements` will contain all implemented protocols? Or is there any particular reason why you decided not to distinguish these types? 

Also there must be cases for templates when it's needed to distinguish inherited type and implemented protocols, though I can not come up with an example right now.

The second thing is that maybe it will make sense not to put in `inherited` types a raw type name for enums, as it is not a case of inheritance per se. And there is already `rawType` for that.

What do you think @krzysztofzablocki ? Will be glad to make a pr for that.

**Comment by krzysztofzablocki** [comment]
Yeah I first wanted that logic to work exactly as you expressed, the reason why I made the 2 be the same set, for now, is because I was concerned about the fact that many base classes / protocols will come from either Apple or 3rd party frameworks and often the Insanity won't know whether they are types or protocols.

We can go back to the logic we both prefer and just have those contain the types we know about, but in that case we should probably add something like `types.based.ClassOrProtocol`, what do you think?

**Comment by ilyapuchka** [comment]
Yes, having base class coming from some other module will make it impossible to guess if it's a type or a protocol 😢 

One very-very far from ideal solution could be to fill `implements` types only when they are added via extension (which is I would say a popular practice) and only put `inherited` type as first type defined in the body of class. For structs and enums we can not require that as they can not inherit at all.

I will search through SourceKitten issues to try to find any insights. Most likely some one already asked that question there.

**Comment by krzysztofzablocki** [comment]
In general my suggestion is for a project to define a fake protocol to markup logic, e.g.
```swift
protocol AutoEquatable {}

// somewhere else
extension MyType: AutoEquatable {}
```

Then in our template you can just go over types implementing it and add `Equatability` conformance.

So let's do this with `inherited` / `implements`:

- `inherited` contains only project defined types and only classes since value types can't inherit
- `implements` contains only project defined protocols
- `based` (or other name if you have one) - contains to all types, including 3rd party ones that we only know the name

Does that sound reasonable? if so, let me know if you'd like to make a PR for it

**Comment by ilyapuchka** [comment]
So if I understand correctly what you suggest is:
add `base` property to replace current `inheriting` property that includes *all* types from which given type derives;
in `inherited` property *for classes only* filter out unknown types
in `implemented` (I would use the same form, either implements/inherits or implemented/inherited, or even keep existing implementing/inheriting) *for any type* filter out unknown protocols

**Comment by krzysztofzablocki** [comment]
Yeah:
- `base` replaces current version, we don't do any filtering on it
- `inherits` contains only types inherting from known project types (that are classes and not value types)
- `implements` contains only types implementing known project protocols

**Comment by krzysztofzablocki** [comment]
I feel like it's reasonable default behavior, and we can always change it later in the future if we decide to

**Comment by ilyapuchka** [comment]
I found that making enums to distinguish inherited types and raw value types is much more complicated, as there are a lot of cases (implicit raw values for all kinds of numeric types and strings, RawRepresentable conformance with explicit typealias or inferred type via `rawValue` property or `init`). Will create a separate issue for that, but not entirely sure it will be possible to achieve.

**Comment by krzysztofzablocki** [comment]
I kind of felt it might be, but happy with the improvement to `inheriting`, `implementing` and `base` as it adds more options to templates :)
